File: GameEngine.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.dylan_cahu.pendu;

/**
 *
 * @author Utilisateur
 */
public class GameEngine {
    
    public  int wrongGuesses = 0;
    public  int maxWrongGuesses = 6;
    public  int goodGuessesCount;
    public  String word = getStartWord();
    public  boolean[] guessedTab = new boolean[word.length()];
    public boolean isWin = false;
    public boolean isLose = false;
    public boolean canPlay = true;
    
    public  String getStartWord() {
        return "test";
    }

    public  boolean checkGuess(char guess) {
        if (word.contains(String.valueOf(guess))) {
            for (int i = 0; i < word.length(); i++) {
                if (word.charAt(i) == guess) {
                    goodGuessesCount++;
                    guessedTab[i] = true;
                }
            }

            if (goodGuessesCount == word.length()) {
                canPlay = false;
                isWin = true;
            }
        } else {
            wrongGuesses++;
            

            if (wrongGuesses == maxWrongGuesses) {
                canPlay = false;
                isLose = false;
            }
        }
        return word.contains(String.valueOf(guess));
    }

    public  String getMaskedWord() {
        StringBuilder maskedWord = new StringBuilder();
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (c == ' ' || guessedTab[i] == true) {
                maskedWord.append(c);
            } else {
                maskedWord.append('*');
            }
        }
        return maskedWord.toString();
    }
}

File: gameEngine.cs
namespace PenduConsole
{
    public class gameEngine
    {
        public int GoodGuessesCount { get; private set; }
        private string word;
        private bool[] guessedLetters;
        public bool CanPlay { get; private set; } = true;
        public bool IsWin { get; private set; } = false;
        public bool IsLose { get; private set; } = false;
        public int ActivePlayerIndex { get; private set; } = 0;
        private Player[] players;

        // Constructor
        public gameEngine(Player[] addPlayers)
        {
            word = getStartWord();
            guessedLetters = new bool[word.Length];
            players = addPlayers;
        }

        public Player getActivePlayer()
        {
            return players[ActivePlayerIndex];
        }

        public bool activePlayerHasLife()
        {
            return getActivePlayer().getLife() > 0;
        }

        private void switchPlayer()
        {
            ActivePlayerIndex++;
            if (ActivePlayerIndex >= players.Length -1)
            {
                ActivePlayerIndex = 0;
            }
            if (activePlayerHasLife() == false)
            {
                int cptPerdant=0;

                for (int i=0 ; i < players.Length -1; i++)
                {
                    if(players[i].getLife() == 0)
                    {
                        cptPerdant++;
                    }
                }
                if (cptPerdant == players.Length-1)
                {
                    Lose();
                }
                else
                {
                    switchPlayer();
                }
            }
        }

        public string getWord() 
        {
            return word;
        }

        public string getStartWord()
        {
            return "test";
        }

        public bool checkGuess(string guess)
        {
            if (guess == "")
            {
                getActivePlayer().decreaseLife();
                switchPlayer();
                return false;
            }
            if (word.Contains(guess.ToLower()[0]))
            {
                for (int i = 0; i < word.Length; i++)
                {
                    if (word[i] == guess.ToLower()[0])
                    {
                        if (guessedLetters[i] == true)
                        {
                            return true;
                        }
                        else
                        {
                            GoodGuessesCount++;
                            guessedLetters[i] = true;
                        }
                    }
                }

                if (GoodGuessesCount == word.Length)
                {
                    Win();
                }
            }
            else
            {
                getActivePlayer().decreaseLife();
            }

            switchPlayer();
            return word.Contains(guess);
        }

        public void Win()
        {
            CanPlay = false;
            IsWin = true;
        }

        public void Lose()
        {
            CanPlay = false;
            IsLose = true;
        }

        public string getMaskedWord()
        {
            string maskedWord = "";
            for (int i = 0; i < word.Length; i++)
            {
                if (word[i] == ' ')
                {
                    maskedWord += " ";
                }
                else
                {
                    if (guessedLetters[i])
                    {
                        maskedWord += word[i];
                    }
                    else
                    {
                        maskedWord += "#";
                    }
                }
            }
            return maskedWord;
        }
    }
}